### PR TITLE
bringing arkode integrator to similar code structure as cvode

### PR DESCRIPTION
**Pull request by hsitaram**
This PR adds the array4 integration capability in arkode, both gpu and cpu. It also makes slight modifications to the RK64 to compile correctly. I will add a similar PR for RK64 for completeness in a later PR. Some things to note:

1) I have moved the AMREX_misc.H outside of cvode so that it can be used by all integrators. There exists 2 of them for CPU and GPU, I am sure it can be consolidated. I will defer to Anne/Marc on this.

2) I have unified some of the run time parameters such as rtol and atol under a common rootname for parmparse. It is set to ode now. May be calling it "chemode" will sound better. Again I defer to Anne/Marc on this.

3) The non array4 version for both GPU/CPU may need to be tested in the ReactEval driver as well. I have made sure arkode 1d array integrator works, not done for CVODE.

4) Moving forward I hope we can change the ".h" files to ".H" and change CVODE_BOXINTG flag to just BOXINTG. I have not done that yet, I hope Ann/Marc agree with me.

### CODE REVIEW

**Comment by jrood-nrel** [comment]
Do you know if this will be faster? I would like to update my plots whenever we know we're making noticeable speed improvements.

File: Testing/Exec/ReactEval_C/GNUmakefile [hunk 1]
```diff
@@ -66,4 +74,6 @@ endif
 Bpack   := ./Make.package
 Blocs   := .
 
+LIBRARIES += -lnvToolsExt
+
```
Comment (drummerdoc) [inline Testing/Exec/ReactEval_C/GNUmakefile]: What is this?  Is this a universal requirement

File: Testing/Exec/ReactEval_C/GNUmakefile [hunk 2]
```diff
@@ -41,16 +41,23 @@ ifeq ($(USE_SUNDIALS_PP), TRUE)
...
   ifeq ($(USE_RK64_PP), TRUE)
   # Activates use of Hari explicit RK
   # will only work if USE_SUNDIALS_PP = FALSE
+<<<<<<< HEAD
   TRANSPORT_CPP = TRUE
   endif
+=======
+  USE_RK64_PP = TRUE
+>>>>>>> 9e95165c70bccecf7f25e11f610c6ee52724a013
 endif
```
Comment (drummerdoc) [inline Testing/Exec/ReactEval_C/GNUmakefile:48]: Need to clean these conflicts out

File: Reactions/Fuego/GPU/arkode/reactor.cpp [hunk 3]
```diff
@@ -51,10 +51,10 @@ int react(const amrex::Box& box,
     realtype time_init, time_out;
 
     void *arkode_mem    = NULL;
-    N_Vector y         = NULL;
+    N_Vector y          = NULL;
 
     NEQ            = NUM_SPECIES+1;
-    NCELLS         = Ncells;
+    NCELLS         = box.numPts();
```
Comment (drummerdoc) [inline Reactions/Fuego/GPU/arkode/reactor.cpp:81]: Is there a potential issue for large box blowing this out (should it be an unsigned long)?

**Review by drummerdoc — APPROVED** [review]

File: Testing/Exec/ReactEval_C/GNUmakefile (hunk 1 above)
Comment (hsitaram) [reply Testing/Exec/ReactEval_C/GNUmakefile]: I can remove this. some remnants from compiling on NREL machine. sorry

File: Reactions/Fuego/GPU/arkode/reactor.cpp (hunk 3 above)
Comment (hsitaram) [reply Reactions/Fuego/GPU/arkode/reactor.cpp:81]: we should probably be safe unless someone is using max_grid_size > 1024. I can change it out to unsigned just to make sure.

Comment (drummerdoc) [reply Reactions/Fuego/GPU/arkode/reactor.cpp:81]: Maybe a better option is to leave it as int for compatibility with SUNDIALS, but to the test in an ALWAYS_ASSERT to make sure it's safe

Comment (hsitaram) [reply Reactions/Fuego/GPU/arkode/reactor.cpp:81]: I have added an AMREX_ASSERT in there now

File: Testing/Exec/ReactEval_C/GNUmakefile (hunk 2 above)
Comment (hsitaram) [reply Testing/Exec/ReactEval_C/GNUmakefile:48]: I think it is clean now.

**Comment by hsitaram** [comment]
> Do you know if this will be faster? I would like to update my plots whenever we know we're making noticeable speed improvements.

@jrood-nrel - I think it will be faster. Some of the cudamemcpys are removed now with allocation done using AMRex_arenas similar to what is done in cvode now.

**Review by drummerdoc — APPROVED** [review]

**Comment by jrood-nrel** [comment]
Awesome. We have time left on Summit. I will benchmark this once it's in.

**Comment by hsitaram** [comment]
Hi Marc, Anne - can this be merged into development so that we can start using this in pelec? Or is it conflicting with your work - Anne

**Review by EnnaDelfen — APPROVED** [review]
looks good to me . I think we should merge. Files changed are all contained in arkode so if bugs there are, they'll be contained.

**Pull request merged**